keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,176 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.
+
+This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.
+
+## Motivation
+
+Today kubernetes doesn't enable providers that are able to safely detach (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Implement in-cluster logic to handle node/control plane partitioning 
+* enable detach for all the storage providers
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+	// SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations, when set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking for detach safety
+	// +optional
+  SafeDetach *bool
+}
+
+```
+
+
+the kube-controller-manager, would need to ensure that eviction looks at the `CSIDriver` and when `.spec.SafeDetach` is set true, forcefully delete pods.
+
+Specifically the pod gc controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+- pods that are backed by a provisioner that support safe detach
+- pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.
```

Comment:
In the existing pod gc controller, there are 3 routines that deletes pods:
- gcTerminated
- gcOrphaned
- gcUnscheduledTerminating

Are you going to modify the existing gcTerminated() to check this `SafeDetach` flag or does this need a new gcXXXX routine?  The KEP didn't describe what triggers the pod gc controller to delete pods other than checking the `SafeDetch` flag and `ReadWriteMany` volume.  Does it go thru the pods and check if pod.Status.Phase is not one of v1.PodPending, v1.PodRunning, and v1.PodUnknown and decide it should be terminated or does it need to check node status?  Please clarify.

Reply:
I don't see the details added ?